### ISSUE TITLE
`FunctionType` is inherently unsafe

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#600]
### Summary

descriptors can be unsafe:
```py
class A: 
    pass

class B(A):
    def __get__(self, *_) -> 1:
        return 1

class C:
    a: A = B()

C.a  # static: A, runtime: 1
```
we can avoid this be reporting an error on the signature of `__get__`: "`__get__`s return type must be assignable to the super classes `__get__` (defaulting to the type of the superclass in it's absense)"

unfortunatly, `FunctionType` commits this sin, and it is an invalid subtype of `Callable`:

```py
class A:
    c: Callable[[object], None] = lambda self: None
A().c  # static: (object) -> None, runtime: () -> None
```

see: https://play.ty.dev/79255fa6-667e-4fa8-b6f2-32225f783bbd

# prior art
basedmypy resolved this problem fairly well by have special cased errors for these assignments to classes for `FunctionType` and `Callable`. the issue is that the corpus of existing code is written under the broken mypy semantics that all `Callable`s are actually `FunctionType`s

perhaps ideally we could just report that `FunctionType` is incompatible with `Callable`:

```py
_: Callable[[], None] = lambda: None  # error: FunctionType is incompatible with Callable, please try again in a few minutes
```

**Comment by carljm** [comment]
Relevant discussion (focused primarily on the `__get__` unsoundness): https://discuss.python.org/t/when-should-we-assume-callable-types-are-method-descriptors/92938

**Comment by carljm** [comment]
I am going to close this as a duplicate of #491 -- it is the same issue, I don't think this is independently actionable.

**Comment by KotlinIsland** [comment]
> I don't think this is independently actionable.

i'll raise the issue regarding descriptors in general
